File: src/lib/runtime.ts
/**
 * Edge mirror of `python/lampforge/runtime.py`.
 *
 * The Python module is the canonical implementation; this file exists because
 * the app's HTTP + MCP endpoints run on the edge. Behaviour must stay
 * one-for-one: same filters, same mode transitions, same experience shape.
 */
import {
  seed,
  type Concept,
  type Decision,
  type Evidence,
  type LearnerState,
  type PedagogyMode,
  type UIComponent,
} from "./lampforge";

export const MODE_NARRATION: Record<PedagogyMode, string> = {
  i_do: "Watch first. I'll build it and narrate every decision as I make it.",
  we_do: "We'll build this together. I'll start it; you decide the next move.",
  you_do: "Your turn. I'll hold the criteria and answer questions, but the moves are yours.",
};

export function listConcepts(opts: { stack?: string; depth?: string; query?: string } = {}): Concept[] {
  let items = seed.concepts as Concept[];
  if (opts.stack) items = items.filter((c) => c.stack === opts.stack);
  if (opts.depth) items = items.filter((c) => c.depth === opts.depth);
  if (opts.query) {
    const q = opts.query.toLowerCase();
    items = items.filter(
      (c) => c.title.toLowerCase().includes(q) || c.summary.toLowerCase().includes(q),
    );
  }
  return items;
}

export const getConcept = (id: string): Concept | undefined => seed.concepts.find((c) => c.id === id);

export const listDecisions = (projectId?: string): Decision[] =>
  projectId ? seed.decisions.filter((d) => d.project_id === projectId) : (seed.decisions as Decision[]);

export const getDecision = (id: string): Decision | undefined => seed.decisions.find((d) => d.id === id);

export const evidenceFor = (decisionId: string): Evidence[] =>
  seed.evidence.filter((e) => e.decision_id === decisionId);

export const getLearner = (learnerId = "demo-learner"): LearnerState => ({
  ...(seed.learner as LearnerState),
  learner_id: learnerId,
});

export function nextMode(state: LearnerState): PedagogyMode {
  const total = state.successful_retrievals + state.failed_retrievals;
  const retrievalRate = total ? state.successful_retrievals / total : 0;
  const struggling = state.hint_requests >= 3 || state.retries >= 4 || state.accuracy < 0.5;
  const secure = state.accuracy >= 0.8 && retrievalRate >= 0.7 && state.hint_requests <= 1;
  const mode = state.mode as PedagogyMode;
  if (struggling) return mode === "we_do" ? "i_do" : "we_do";
  if (secure) return mode === "we_do" ? "you_do" : "we_do";
  return mode;
}

export function explain(conceptId: string, mode: PedagogyMode = "we_do") {
  const concept = getConcept(conceptId);
  if (!concept) return null;
  const body =
    mode === "i_do"
      ? `${concept.summary} Here is the worked form: ${concept.code_example ?? "see the artefact"}`
      : mode === "we_do"
        ? `${concept.summary} Given that, what would you change first?`
        : `Before I say anything: describe ${concept.title.toLowerCase()} in your own words, then I'll respond to what you said.`;
  return {
    concept,
    mode,
    narration: MODE_NARRATION[mode],
    body,
    watch_for: concept.common_errors,
    vocabulary: concept.vocabulary,
  };
}

export function assessEvidence(conceptIds: string[], body: string, learnerId = "demo-learner") {
  const text = body.toLowerCase();
  let hits = 0;
  let total = 0;
  for (const cid of conceptIds) {
    const concept = getConcept(cid);
    if (!concept) continue;
    for (const word of concept.vocabulary) {
      total += 1;
      if (text.includes(word.toLowerCase())) hits += 1;
    }
  }
  const ratio = total ? hits / total : 0;
  const justified = ["because", "so that", "instead of", "trade-off", "rather than"].some((w) =>
    text.includes(w),
  );
  const verdict = ratio >= 0.4 && justified ? "secure" : ratio >= 0.2 || justified ? "developing" : "not_evidenced";
  return {
    concept_ids: conceptIds,
    verdict,
    reasoning:
      `Used ${hits} of ${total} target vocabulary items` +
      (justified ? "; gave a reason for the choice." : "; stated the choice without a reason."),
    next_mode: nextMode(getLearner(learnerId)),
    suggested_decision_id:
      seed.decisions.find((d) => conceptIds.some((c) => d.concept_ids.includes(c)))?.id ?? null,
  };
}

const PHP_BY_COMPONENT: Record<string, string> = {
  "ui-booking-form": `<?php
$errors = [];
if ($_SERVER['REQUEST_METHOD'] === 'POST') {
    $name  = trim($_POST['name'] ?? '');
    $email = trim($_POST['email'] ?? '');
    if ($name === '') { $errors[] = 'Name is required'; }
    if (!filter_var($email, FILTER_VALIDATE_EMAIL)) { $errors[] = 'Invalid email format'; }
    if (!$errors) {
        $stmt = $pdo->prepare('INSERT INTO bookings (customer_id, slot_id) VALUES (?, ?)');
        $stmt->execute([$customerId, $slotId]);
    }
}
`,
  "ui-booking-table": `<?php
$stmt = $pdo->prepare(
    'SELECT b.id, c.full_name, s.starts_at, b.status
     FROM bookings b
     JOIN customers c ON c.id = b.customer_id
     JOIN slots s ON s.id = b.slot_id
     ORDER BY s.starts_at'
);
$stmt->execute();
$rows = $stmt->fetchAll(PDO::FETCH_ASSOC);
`,
  "ui-validation-errors": `<?php
// Re-render the form with the submitted values preserved.
$old = fn(string $k) => htmlspecialchars($_POST[$k] ?? '', ENT_QUOTES);
`,
};

const SQL_BY_COMPONENT: Record<string, string> = {
  "ui-booking-form": `INSERT INTO bookings (customer_id, slot_id, status)
VALUES (?, ?, 'pending');`,
  "ui-booking-table": `SELECT b.id, c.full_name, s.starts_at, b.status
FROM bookings b
JOIN customers c ON c.id = b.customer_id
JOIN slots s ON s.id = b.slot_id
ORDER BY s.starts_at;`,
};

function pickComponent(intent: string, conceptIds: string[]): UIComponent {
  const text = intent.toLowerCase();
  const byId = (id: string) => seed.components.find((c) => c.id === id) as UIComponent;
  if (["error", "invalid", "validation", "fail"].some((w) => text.includes(w)))
    return byId("ui-validation-errors");
  if (["table", "list", "bookings", "read", "report"].some((w) => text.includes(w)))
    return byId("ui-booking-table");
  if (conceptIds.includes("crud")) return byId("ui-booking-table");
  return byId("ui-booking-form");
}

export type ExperienceRequestInput = {
  intent: string;
  mode?: PedagogyMode;
  concept_ids?: string[];
  learner_id?: string;
};

export function createExperience(request: ExperienceRequestInput) {
  const mode = request.mode ?? "we_do";
  const component = pickComponent(request.intent, request.concept_ids ?? []);
  const conceptIds = request.concept_ids?.length ? request.concept_ids : component.concept_ids;
  const decision = seed.decisions.find((d) => conceptIds.some((c) => d.concept_ids.includes(c)));

  const followUps =
    mode === "i_do"
      ? [
          `Why is ${component.bootstrap_classes[0]} the outer class here?`,
          "Which line would break first if the input were unlabelled?",
        ]
      : mode === "we_do"
        ? [
            "What should happen when the slot is already taken?",
            "Where would you put the server-side check, and why there?",
          ]
        : [
            "Rebuild this from the schema alone, then compare.",
            "State the consequence you accepted by choosing this layout.",
          ];

  let hash = 0;
  for (const ch of request.intent) hash = (hash * 31 + ch.charCodeAt(0)) >>> 0;

  return {
    id: `exp-${String(hash % 100000).padStart(5, "0")}`,
    intent: {
      verb: "build",
      subject: component.id,
      mode,
      surface: "swagger",
      raw: request.intent,
    },
    mode,
    narration: MODE_NARRATION[mode],
    component,
    php: PHP_BY_COMPONENT[component.id] ?? null,
    sql: SQL_BY_COMPONENT[component.id] ?? null,
    concept_ids: conceptIds,
    decision_id: decision?.id ?? null,
    follow_up_questions: followUps,
  };
}

export type Experience = ReturnType<typeof createExperience>;

File: python/lampforge/runtime.py
"""Pure runtime logic shared by the FastAPI app and the MCP server.

Deliberately free of framework imports so both surfaces are thin projections.
"""

from __future__ import annotations

from typing import List, Optional

from .models import (
    Concept,
    Decision,
    Depth,
    Evidence,
    Experience,
    ExperienceRequest,
    Intent,
    LearnerState,
    PedagogyMode,
    Stack,
    UIComponent,
)
from .seed import build_seed
from .tool_io import AssessmentResult

SEED = build_seed()


# --------------------------------------------------------------------------
# Reads
# --------------------------------------------------------------------------


def list_concepts(
    stack: Optional[Stack] = None,
    depth: Optional[Depth] = None,
    query: Optional[str] = None,
) -> List[Concept]:
    items = SEED.concepts
    if stack:
        items = [c for c in items if c.stack == stack]
    if depth:
        items = [c for c in items if c.depth == depth]
    if query:
        q = query.lower()
        items = [c for c in items if q in c.title.lower() or q in c.summary.lower()]
    return items


def get_concept(concept_id: str) -> Optional[Concept]:
    return next((c for c in SEED.concepts if c.id == concept_id), None)


def list_decisions(project_id: Optional[str] = None) -> List[Decision]:
    if project_id:
        return [d for d in SEED.decisions if d.project_id == project_id]
    return SEED.decisions


def get_decision(decision_id: str) -> Optional[Decision]:
    return next((d for d in SEED.decisions if d.id == decision_id), None)


def evidence_for(decision_id: str) -> List[Evidence]:
    return [e for e in SEED.evidence if e.decision_id == decision_id]


def get_learner(learner_id: str = "demo-learner") -> LearnerState:
    return SEED.learner.model_copy(update={"learner_id": learner_id})


# --------------------------------------------------------------------------
# Pedagogy
# --------------------------------------------------------------------------

_MODE_NARRATION = {
    PedagogyMode.I_DO: "Watch first. I'll build it and narrate every decision as I make it.",
    PedagogyMode.WE_DO: "We'll build this together. I'll start it; you decide the next move.",
    PedagogyMode.YOU_DO: "Your turn. I'll hold the criteria and answer questions, but the moves are yours.",
}


def next_mode(state: LearnerState) -> PedagogyMode:
    """Gradual release, driven by the interaction signals."""
    retrieval_total = state.successful_retrievals + state.failed_retrievals
    retrieval_rate = state.successful_retrievals / retrieval_total if retrieval_total else 0.0
    struggling = state.hint_requests >= 3 or state.retries >= 4 or state.accuracy < 0.5
    secure = state.accuracy >= 0.8 and retrieval_rate >= 0.7 and state.hint_requests <= 1

    if struggling:
        return PedagogyMode.I_DO if state.mode == PedagogyMode.WE_DO else PedagogyMode.WE_DO
    if secure:
        return PedagogyMode.YOU_DO if state.mode == PedagogyMode.WE_DO else PedagogyMode.WE_DO
    return state.mode


def explain(concept_id: str, mode: PedagogyMode = PedagogyMode.WE_DO) -> dict:
    concept = get_concept(concept_id)
    if concept is None:
        raise KeyError(concept_id)
    if mode == PedagogyMode.I_DO:
        body = f"{concept.summary} Here is the worked form: {concept.code_example or 'see the artefact'}"
    elif mode == PedagogyMode.WE_DO:
        body = f"{concept.summary} Given that, what would you change first?"
    else:
        body = f"Before I say anything: describe {concept.title.lower()} in your own words, then I'll respond to what you said."
    return {
        "concept": concept.model_dump(),
        "mode": mode.value,
        "narration": _MODE_NARRATION[mode],
        "body": body,
        "watch_for": concept.common_errors,
        "vocabulary": concept.vocabulary,
    }


def assess_evidence(concept_ids: List[str], body: str, learner_id: str = "demo-learner") -> AssessmentResult:
    """Heuristic assessment. The chat runtime replaces this with a model call."""
    text = body.lower()
    hits = 0
    total = 0
    for cid in concept_ids:
        concept = get_concept(cid)
        if not concept:
            continue
        for word in concept.vocabulary:
            total += 1
            if word.lower() in text:
                hits += 1
    ratio = hits / total if total else 0.0
    justified = any(w in text for w in ("because", "so that", "instead of", "trade-off", "rather than"))

    if ratio >= 0.4 and justified:
        verdict = "secure"
    elif ratio >= 0.2 or justified:
        verdict = "developing"
    else:
        verdict = "not_evidenced"

    reasoning = (
        f"Used {hits} of {total} target vocabulary items"
        + ("; gave a reason for the choice." if justified else "; stated the choice without a reason.")
    )
    state = get_learner(learner_id)
    suggested = next(
        (d.id for d in SEED.decisions if any(c in d.concept_ids for c in concept_ids)),
        None,
    )
    return AssessmentResult(
        concept_ids=concept_ids,
        verdict=verdict,
        reasoning=reasoning,
        next_mode=next_mode(state),
        suggested_decision_id=suggested,
    )


# --------------------------------------------------------------------------
# Experiences — the multi-representation learning object
# --------------------------------------------------------------------------

_PHP_BY_COMPONENT = {
    "ui-booking-form": (
        "<?php\n"
        "$errors = [];\n"
        "if ($_SERVER['REQUEST_METHOD'] === 'POST') {\n"
        "    $name  = trim($_POST['name'] ?? '');\n"
        "    $email = trim($_POST['email'] ?? '');\n"
        "    if ($name === '') { $errors[] = 'Name is required'; }\n"
        "    if (!filter_var($email, FILTER_VALIDATE_EMAIL)) { $errors[] = 'Invalid email format'; }\n"
        "    if (!$errors) {\n"
        "        $stmt = $pdo->prepare('INSERT INTO bookings (customer_id, slot_id) VALUES (?, ?)');\n"
        "        $stmt->execute([$customerId, $slotId]);\n"
        "    }\n"
        "}\n"
    ),
    "ui-booking-table": (
        "<?php\n"
        "$stmt = $pdo->prepare(\n"
        "    'SELECT b.id, c.full_name, s.starts_at, b.status\n"
        "     FROM bookings b\n"
        "     JOIN customers c ON c.id = b.customer_id\n"
        "     JOIN slots s ON s.id = b.slot_id\n"
        "     ORDER BY s.starts_at'\n"
        ");\n"
        "$stmt->execute();\n"
        "$rows = $stmt->fetchAll(PDO::FETCH_ASSOC);\n"
    ),
    "ui-validation-errors": (
        "<?php\n"
        "// Re-render the form with the submitted values preserved.\n"
        "$old = fn(string $k) => htmlspecialchars($_POST[$k] ?? '', ENT_QUOTES);\n"
    ),
}

_SQL_BY_COMPONENT = {
    "ui-booking-form": (
        "INSERT INTO bookings (customer_id, slot_id, status)\n"
        "VALUES (?, ?, 'pending');"
    ),
    "ui-booking-table": (
        "SELECT b.id, c.full_name, s.starts_at, b.status\n"
        "FROM bookings b\n"
        "JOIN customers c ON c.id = b.customer_id\n"
        "JOIN slots s ON s.id = b.slot_id\n"
        "ORDER BY s.starts_at;"
    ),
}


def _pick_component(intent: str, concept_ids: List[str]) -> UIComponent:
    text = intent.lower()
    if any(w in text for w in ("error", "invalid", "validation", "fail")):
        return next(c for c in SEED.components if c.id == "ui-validation-errors")
    if any(w in text for w in ("table", "list", "bookings", "read", "report")):
        return next(c for c in SEED.components if c.id == "ui-booking-table")
    if concept_ids and "crud" in concept_ids:
        return next(c for c in SEED.components if c.id == "ui-booking-table")
    return next(c for c in SEED.components if c.id == "ui-booking-form")


def create_experience(request: ExperienceRequest) -> Experience:
    component = _pick_component(request.intent, request.concept_ids)
    concept_ids = request.concept_ids or component.concept_ids
    decision = next(
        (d for d in SEED.decisions if any(c in d.concept_ids for c in concept_ids)),
        None,
    )
    if request.mode == PedagogyMode.I_DO:
        follow_ups = [
            f"Why is {component.bootstrap_classes[0]} the outer class here?",
            "Which line would break first if the input were unlabelled?",
        ]
    elif request.mode == PedagogyMode.WE_DO:
        follow_ups = [
            "What should happen when the slot is already taken?",
            "Where would you put the server-side check, and why there?",
        ]
    else:
        follow_ups = [
            "Rebuild this from the schema alone, then compare.",
            "State the consequence you accepted by choosing this layout.",
        ]

    return Experience(
        id=f"exp-{abs(hash(request.intent)) % 100000:05d}",
        intent=Intent(
            verb="build",
            subject=component.id,
            mode=request.mode,
            surface="swagger",
            raw=request.intent,
        ),
        mode=request.mode,
        narration=_MODE_NARRATION[request.mode],
        component=component,
        php=_PHP_BY_COMPONENT.get(component.id),
        sql=_SQL_BY_COMPONENT.get(component.id),
        concept_ids=concept_ids,
        decision_id=decision.id if decision else None,
        follow_up_questions=follow_ups,
    )
